### PR TITLE
Fix: Loss of trailing digits when adding outfit

### PR DESCRIPTION
**Pull request by oo13**
**Bugfix:** This PR addresses issue #3639 and #3645.

## Fix Details
This PR addresses an issue that is caused by an accumulation of a loss of trailing digits when adding an outfit.

#3645 could solve the issue by another way; this PR needs more calculation amount, but is more proper and removes a mysterious parameter EPS.

## Testing Done
1. Load a pilot Sell Issue.
2. Sell all Laser Rifles.
3. Sell all Grenades.
4. Try to sell all Nerve Gases.

Current master cannot sell a last Nerve Gas, but this PR can sell all Nerve Gases.

## Save File
Test file provided by @KiLEdEnNis (same as #3645): [sell_issue.txt](https://github.com/endless-sky/endless-sky/files/1722054/sell_issue.txt)

## Performance Impact
This PR affects a performance only when Adding/Removing outfits and the impact is not noticeable.

### CODE REVIEW

File: source/Outfit.cpp [hunk 1]
```diff
@@ -88,6 +87,31 @@ namespace {
 				thisMap.erase(it.first);
 		}
 	}
+	
+	// Summation that avoids a loss of trailing digits as possible.
```
Comment (tatami4) [inline source/Outfit.cpp:91]: Shouldn't this be `// Summation that avoids loss of trailing digits as much as possible.`

File: source/Outfit.cpp [hunk 2]
```diff
@@ -88,6 +87,31 @@ namespace {
...
+		vector<double> partialSums;
+		partialSums.reserve(partialTerms.size());
+		for(const auto &it : partialTerms)
+			partialSums.push_back(it.first * it.second);
+		sort(partialSums.begin(), partialSums.end(), [&](double a, double b){ return fabs(a) < fabs(b); });
+		return accumulate(partialSums.begin(), partialSums.end(), 0.);
+	}
+	
+	// Add value * count to an attribute.
+	// This function doesn't modify attribute if valud * count is zero.
```
Comment (tatami4) [inline source/Outfit.cpp:103]: typo: `if valud * count`

**Comment by oo13** [comment]
@tatami4 Thank you for suggestions. I posted the PR applied your comments.

File: source/Outfit.cpp [hunk 3]
```diff
@@ -88,6 +87,31 @@ namespace {
...
 	}
+	
+	// Summation that avoids loss of trailing digits as much as possible.
+	double Sum(const map<double, int> &partialTerms)
+	{
+		vector<double> partialSums;
+		partialSums.reserve(partialTerms.size());
+		for(const auto &it : partialTerms)
+			partialSums.push_back(it.first * it.second);
+		sort(partialSums.begin(), partialSums.end(), [&](double a, double b){ return fabs(a) < fabs(b); });
```
Comment (petervdmeer) [inline source/Outfit.cpp:98]: If we want to avoid loss of trailing digits as much as possible, then I would expect the smallest numbers to be accumulated first before the large numbers. Because the loss mostly happens when small numbers are added to large numbers.

File: source/Outfit.h [hunk 4]
```diff
@@ -105,6 +105,7 @@ class Outfit : public Weapon {
 	std::vector<std::string> licenses;
 	
 	Dictionary attributes;
+	std::map<std::string, std::map<double, int>> partialTerms;
```
Comment (petervdmeer) [inline source/Outfit.h:108]: The code here looks quite elegant, but please add a few comments here on how the partialTerms help to avoid the rounding errors (because I expect that it is a bit hard to follow for people that just encounter this code later out of the context of this PR).

File: source/Outfit.cpp [hunk 5]
```diff
@@ -88,6 +87,31 @@ namespace {
...
+		return accumulate(partialSums.begin(), partialSums.end(), 0.);
+	}
+	
+	// Add value * count to an attribute.
+	// This function doesn't modify attribute if value * count is zero.
+	void AddToAttribute(double &attribute, map<double, int> &partialTerms, double value, int count)
```
Comment (petervdmeer) [inline source/Outfit.cpp:104]: Do all additions and removals always use the proper `base value`, e.g. if we have situations where an attribute has `+0.5` and `2* -0.25`, then the magic added here wouldn't work well.

File: source/Outfit.cpp [hunk 6]
```diff
@@ -88,6 +87,31 @@ namespace {
 				thisMap.erase(it.first);
 		}
 	}
+	
+	// Summation that avoids loss of trailing digits as much as possible.
+	double Sum(const map<double, int> &partialTerms)
```
Comment (petervdmeer) [inline source/Outfit.cpp:92]: The memory usage for attributes increases quite a bit using this solution; we now get an entry in the partialTerms map for every fraction that was added to the attribute. This will happen for every ship, including ships where the outfits are typically not changed.
If we go this route, then there might be another solution that uses an equal amount of memory, but that computes faster and that is to use very high precision numbers instead of the partial terms.

Another consideration is that we only need this very high precision when we are actually changing ship outfits, and the raw data to generate the high precision already seems to be stored in the outfit-counts on a ship. We should be able to generate the high precision data on-the-fly when we need them during Outfitting (Thus reducing the overal memory usage at the costs of additional CPU time when we are starting to change Outfits.)

File: source/Outfit.cpp (hunk 5 above)
Comment (oo13) [reply source/Outfit.cpp:104]: I cannot remember it exactly, but I guess the only intention is to avoid infinitely increasing the size of "partialTerms."

**Comment by oo13** [comment]
If you want to fix the root cause of the precision problem, I recommend that you design it with the fixed-point and implement it with the double precision. It is very easy and less modification than you probably expect.

**Comment by petervdmeer** [comment]
> If you want to fix the root cause of the precision problem, I recommend that you design it with the fixed-point and implement it with the double precision. It is very easy and less modification than you probably expect.

Yes, doing the actual calculations in base-2 fixed-point would likely resolve the problem completely. (And keeping the doubles "at flight time" should result in zero/no performance impact.) Would you like to give that a try and (create a new) PR for) this?

**Comment by oo13** [comment]
I have no plan of a PR.
You need to determine the dynamic ranges and the LSB positions for the maximum number of the outfits and all attributes before implementing it, anyway.

**Comment by petervdmeer** [comment]
@oo13 Good to know! We will take your advise and this PR as inspiration to address the issue.
